doc/s1kd-highlight.1
.\" Automatically generated by Pandoc 1.19.2.1
.\"
.TH "s1kd\-highlight" "1" "2018\-04\-11" "" "General Commands Manual"
.hy
.SH NAME
.PP
s1kd\-highlight \- Syntax highlighting for S1000D verbatimText
.SH SYNOPSIS
.IP
.nf
\f[C]
s1kd\-highlight\ [options]\ [<dmodule>...]
\f[]
.fi
.SH DESCRIPTION
.PP
The \f[I]s1kd\-highlight\f[] tool adds syntax highlighting to program
listings in a data module.
By default this is accomplished by wrapping detected syntax in XSL\-FO
inline elements with the proper namespace to allow a stylesheet to pass
them through to the final FO processor.
The actual elements used can be customized, however.
.PP
To enable highlighting on a verbatim text element, include the
processing instruction \f[C]<?language\ ...?>\f[] in the element, where
\f[C]\&...\f[] is the name of the language.
.SH OPTIONS
.TP
.B \-c <classes>
Use a custom class definitions file.
.RS
.RE
.TP
.B \-f
Overwrite input data module(s) instead of outputting to stdout.
.RS
.RE
.TP
.B \-h \-?
Show help/usage message.
.RS
.RE
.TP
.B \-s <syntax>
Use a custom syntax definitions file.
.RS
.RE
.SH CLASS FILE FORMAT
.PP
The following describes the format of the custom class file specified
with \-c.
.SS Classes
.PP
\f[I]Markup element:\f[] <\f[C]classes\f[]>
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
None
.PP
\f[I]Child elements:\f[]
.IP \[bu] 2
<\f[C]class\f[]>
.SS Class
.PP
Represents a type of syntax and how it should be highlighted.
This element can also occur within the \f[C]syntax\f[] element or within
a particular \f[C]language\f[] element, in which case it is specific to
that language.
.PP
\f[I]Markup element:\f[] <\f[C]class\f[]>
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
\f[C]id\f[], the identifier of the class.
.PP
\f[I]Child elements:\f[]
.PP
The element \f[C]class\f[] contains one child element of any kind, which
any matching syntax will be wrapped in to.
.SS Example custom classes file
.IP
.nf
\f[C]
<classes\ xmlns:fo="http://www.w3.org/1999/XSL/Format">
<class\ id="type">
<fo:inline\ color="green"/>
</class>
<class\ id="control">
<fo:inline\ color="blue"/>
</class>
<class\ id="string">
<fo:inline\ color="red"/>
</class>
<class\ id="comment">
<fo:inline\ color="pink"/>
</class>
</classes>
\f[]
.fi
.SH SYNTAX FILE FORMAT
.PP
The following describes the format of the custom syntax file specified
with \-s.
.SS Syntax
.PP
\f[I]Markup element:\f[] <\f[C]syntax\f[]>
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
None
.PP
\f[I]Child elements:\f[]
.IP \[bu] 2
\f[C]class\f[]
.IP \[bu] 2
\f[C]language\f[]
.SS Language
.PP
Describes the syntax of a particular language.
.PP
\f[I]Markup element:\f[] <\f[C]language\f[]>
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
\f[C]name\f[], the identifier of the language.
.IP \[bu] 2
\f[C]caseInsensitive\f[], indicates whether keywords are
case\-insensitive in this language, with one of the following values:
.RS 2
.IP \[bu] 2
\f[C]"no"\f[] \- Keywords are case\-sensitive (default)
.IP \[bu] 2
\f[C]"yes"\f[] \- Keywords are case\-insensitive
.RE
.PP
\f[I]Child elements:\f[]
.IP \[bu] 2
<\f[C]class\f[]>
.IP \[bu] 2
<\f[C]area\f[]>
.IP \[bu] 2
<\f[C]keyword\f[]>
.SS Area
.PP
Matches a section of delimited text, such as strings, comments, etc.
.PP
\f[I]Markup element:\f[] <\f[C]area\f[]>
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
\f[C]start\f[], the opening delimiter.
.IP \[bu] 2
\f[C]end\f[], the closing delimiter.
.IP \[bu] 2
\f[C]class\f[], reference to the \f[C]class\f[] element to use for this
area.
.PP
\f[I]Child elements:\f[]
.PP
If attribute \f[C]class\f[] is not used, this element can contain one
element of any kind, in which the text matching the area will be
wrapped.
.SS Keyword
.PP
Matches a particular keyword.
.PP
\f[I]Markup element:\f[] <\f[C]keyword\f[]>
.PP
\f[I]Attributes:\f[]
.IP \[bu] 2
\f[C]match\f[], the keyword to match.
.IP \[bu] 2
\f[C]class\f[], reference to the \f[C]class\f[] element to use for this
keyword.
.PP
\f[I]Child elements:\f[]
.PP
If attribute \f[C]class\f[] is not used, this element can contain one
element of any kind, in which the text matching the keyword will be
wrapped.
.SS Example custom syntax file
.IP
.nf
\f[C]
<syntax>
<language\ name="c">
<area\ start="&quot;"\ end="&quot;"\ class="string"/>
<area\ start="/*"\ end="*/"\ class="comment"/>
<keyword\ match="if"\ class="control"/>
<keyword\ match="else"\ class="control"/>
<keyword\ match="int"\ class="type"/>
<keyword\ match="char"\ class="type"/>
</language>
</syntax>
\f[]
.fi
.SH AUTHORS
khzae.net.


s1kd-highlight.c
#include <stdio.h>
#include <stdbool.h>
#include <unistd.h>
#include <string.h>
#include <libxml/tree.h>
#include <libxml/xpath.h>
#include "languages.h"

#define PROG_NAME "s1kd-highlight"

#define PRE_KEYWORD_DELIM BAD_CAST " (\n"
#define POST_KEYWORD_DELIM BAD_CAST " .,);\n"

#define VERBATIM_TEXT_XPATH "//verbatimText[processing-instruction('language')='%s']/text()"

xmlNodePtr first_xpath_node(xmlDocPtr doc, xmlNodePtr node, const xmlChar *expr)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;
	xmlNodePtr first;

	ctx = xmlXPathNewContext(doc ? doc : node->doc);
	ctx->node = node;

	obj = xmlXPathEvalExpression(expr, ctx);

	first = xmlXPathNodeSetIsEmpty(obj->nodesetval) ? NULL : obj->nodesetval->nodeTab[0];

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);

	return first;
}

bool is_keyword(xmlChar *content, int content_len, int i, const xmlChar *keyword, int keyword_len, bool ignorecase)
{
	bool is;
	xmlChar s, e;
	int (*cmp)(const xmlChar *, const xmlChar *, int);

	cmp = ignorecase ? xmlStrncasecmp : xmlStrncmp;

	s = i == 0 ? ' ' : content[i - 1];
	e = i + keyword_len >= content_len - 1 ? ' ' : content[i + keyword_len];

	is = xmlStrchr(PRE_KEYWORD_DELIM, s) &&
	     cmp(content + i, keyword, keyword_len) == 0 &&
	     xmlStrchr(POST_KEYWORD_DELIM, e);

	return is;
}

void highlight_keyword_in_node(xmlNodePtr node, const xmlChar *keyword, xmlNodePtr tag, bool ignorecase)
{
	xmlChar *content;
	int keyword_len, content_len;
	int i;

	content = xmlNodeGetContent(node);
	content_len = xmlStrlen(content);

	keyword_len = xmlStrlen(keyword);

	i = 0;
	while (i + keyword_len <= content_len) {
		if (is_keyword(content, content_len, i, keyword, keyword_len, ignorecase)) {
			xmlChar *s1 = xmlStrndup(content, i);
			xmlChar *s2 = xmlStrsub(content, i + keyword_len, content_len - (i + keyword_len));
			xmlChar *s3 = xmlStrsub(content, i, keyword_len);
			xmlNodePtr elem;

			xmlFree(content);

			xmlNodeSetContent(node, s1);
			xmlFree(s1);

			elem = xmlAddNextSibling(node, xmlCopyNode(tag, 1));
			xmlNodeSetContent(elem, s3);
			xmlFree(s3);

			node = xmlAddNextSibling(elem, xmlNewText(s2));

			content = s2;
			content_len = xmlStrlen(s2);
			i = 0;
		} else {
			++i;
		}
	}

	xmlFree(content);
}

void highlight_area_in_node(xmlNodePtr node, const xmlChar *start, const xmlChar *end, xmlNodePtr tag, bool ignorecase)
{
	xmlChar *content;
	int i, slen, elen;
	int (*cmp)(const xmlChar *, const xmlChar *, int);
	const xmlChar *(*sub)(const xmlChar *, const xmlChar *);

	content = xmlNodeGetContent(node);
	slen = xmlStrlen(start);
	elen = xmlStrlen(end);

	cmp = ignorecase ? xmlStrncasecmp : xmlStrncmp;
	sub = ignorecase ? xmlStrcasestr : xmlStrstr;

	for (i = 0; content[i]; ++i) {
		if (cmp(content + i, start, slen) == 0) {
			const xmlChar *e;
			int len;
			xmlChar *s1, *s2, *s3;
			xmlNodePtr elem;

			e = sub(content + i + 1, end);
			if (!e) {
				e = content + (xmlStrlen(content) - 1);
			}

			len = e - (content + i) + elen;

			s1 = xmlStrndup(content, i);
			s2 = xmlStrndup(content + i, len);
			s3 = xmlStrdup(content + i + len);

			xmlFree(content);

			xmlNodeSetContent(node, s1);
			xmlFree(s1);

			elem = xmlAddNextSibling(node, xmlCopyNode(tag, 1));
			xmlNodeSetContent(elem, s2);
			xmlFree(s2);

			node = xmlAddNextSibling(elem, xmlNewText(s3));

			content = s3;

			i = 0;
		}
	}

	xmlFree(content);
}

void highlight_keyword_in_nodes(xmlNodeSetPtr nodes, const xmlChar *keyword, xmlNodePtr tag, bool ignorecase)
{
	int i;
	for (i = 0; i < nodes->nodeNr; ++i) {
		highlight_keyword_in_node(nodes->nodeTab[i], keyword, tag, ignorecase);
	}
}

void highlight_area_in_nodes(xmlNodeSetPtr nodes, const xmlChar *start, const xmlChar *end, xmlNodePtr tag, bool ignorecase)
{
	int i;
	for (i = 0; i < nodes->nodeNr; ++i) {
		highlight_area_in_node(nodes->nodeTab[i], start, end, tag, ignorecase);
	}
}

void highlight_keyword_in_doc(xmlDocPtr doc, const xmlChar *lang, const xmlChar *keyword, xmlNodePtr tag, bool ignorecase)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;
	
	xmlChar xpath[256];

	xmlStrPrintf(xpath, 256, VERBATIM_TEXT_XPATH, lang);

	ctx = xmlXPathNewContext(doc);
	obj = xmlXPathEvalExpression(xpath, ctx);

	if (!xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
		highlight_keyword_in_nodes(obj->nodesetval, keyword, tag, ignorecase);
	}

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);
}

void highlight_area_in_doc(xmlDocPtr doc, const xmlChar *lang, const xmlChar *start, const xmlChar *end, xmlNodePtr tag, bool ignorecase)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;

	xmlChar xpath[256];

	xmlStrPrintf(xpath, 256, VERBATIM_TEXT_XPATH, lang);

	ctx = xmlXPathNewContext(doc);
	obj = xmlXPathEvalExpression(xpath, ctx);

	if (!xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
		highlight_area_in_nodes(obj->nodesetval, start, end, tag, ignorecase);
	}

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);
}

xmlNodePtr get_class(const xmlChar *class, xmlDocPtr classes, xmlDocPtr syntax)
{
	xmlNodePtr node;
	xmlChar xpath[256];

	xmlStrPrintf(xpath, 256, "//class[@id='%s']", class);

	node = first_xpath_node(classes, NULL, xpath);

	if (!node) {
		node = first_xpath_node(syntax, NULL, xpath);
	}

	return node;
}

void highlight_keyword_node_in_doc(xmlDocPtr doc, xmlDocPtr syntax, xmlDocPtr classes, const xmlChar *lang, xmlNodePtr node, bool ignorecase)
{
	xmlChar *keyword, *class;
	xmlNodePtr tag;

	keyword = xmlGetProp(node, BAD_CAST "match");
	class   = xmlGetProp(node, BAD_CAST "class");

	if (class) {
		tag = xmlFirstElementChild(get_class(class, classes, syntax));
	} else {
		tag = xmlFirstElementChild(node);
	}

	if (tag) {
		highlight_keyword_in_doc(doc, lang, keyword, tag, ignorecase);
	}

	xmlFree(keyword);
	xmlFree(class);
}

void highlight_area_node_in_doc(xmlDocPtr doc, xmlDocPtr syntax, xmlDocPtr classes, const xmlChar *lang, xmlNodePtr node, bool ignorecase)
{
	xmlChar *start, *end, *class;
	xmlNodePtr tag;

	start = xmlGetProp(node, BAD_CAST "start");
	end   = xmlGetProp(node, BAD_CAST "end");
	class = xmlGetProp(node, BAD_CAST "class");

	if (class) {
		tag = xmlFirstElementChild(get_class(class, classes, syntax));
	} else {
		tag = xmlFirstElementChild(node);
	}

	if (tag) {
		highlight_area_in_doc(doc, lang, start, end, tag, ignorecase);
	}

	xmlFree(start);
	xmlFree(end);
	xmlFree(class);
}

void highlight_language_in_doc(xmlDocPtr doc, xmlDocPtr syntax,
	xmlDocPtr classes, xmlNodePtr language)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;
	xmlChar *lang, *case_insen;
	bool ignorecase = false;

	lang = xmlGetProp(language, BAD_CAST "name");
	case_insen = xmlGetProp(language, BAD_CAST "caseInsensitive");

	if (case_insen) {
		ignorecase = xmlStrcmp(case_insen, BAD_CAST "yes") == 0;
	}

	xmlFree(case_insen);

	ctx = xmlXPathNewContext(syntax);
	ctx->node = language;

	obj = xmlXPathEvalExpression(BAD_CAST "area", ctx);

	if (!xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
		int i;
		for (i = 0; i < obj->nodesetval->nodeNr; ++i) {
			highlight_area_node_in_doc(doc, syntax, classes, lang, obj->nodesetval->nodeTab[i], ignorecase);
		}
	}

	xmlXPathFreeObject(obj);

	obj = xmlXPathEvalExpression(BAD_CAST "keyword", ctx);

	if (!xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
		int i;
		for (i = 0; i < obj->nodesetval->nodeNr; ++i) {
			highlight_keyword_node_in_doc(doc, syntax, classes, lang, obj->nodesetval->nodeTab[i], ignorecase);
		}
	}

	xmlFree(lang);

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);
}

void highlight_syntax_in_doc(xmlDocPtr doc, xmlDocPtr syntax, xmlDocPtr classes)
{
	xmlXPathContextPtr ctx;
	xmlXPathObjectPtr obj;

	ctx = xmlXPathNewContext(syntax);
	obj = xmlXPathEvalExpression(BAD_CAST "//language", ctx);

	if (!xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
		int i;
		for (i = 0; i < obj->nodesetval->nodeNr; ++i) {
			highlight_language_in_doc(doc, syntax, classes,
				obj->nodesetval->nodeTab[i]);
		}
	}

	xmlXPathFreeObject(obj);
	xmlXPathFreeContext(ctx);
}

void highlight_syntax_in_file(const char *fname, const char *syntax, const char *classes, bool overwrite)
{
	xmlDocPtr doc;
	xmlDocPtr syndoc;
	xmlDocPtr classdoc;

	if (syntax) {
		syndoc = xmlReadFile(syntax, NULL, 0);
	} else {
		syndoc = xmlReadMemory((const char *) syntax_xml, syntax_xml_len, NULL, NULL, 0);
	}

	if (classes) {
		classdoc = xmlReadFile(classes, NULL, 0);
	} else {
		classdoc = xmlReadMemory((const char *) classes_xml,
			classes_xml_len, NULL, NULL, 0);
	}

	doc = xmlReadFile(fname, NULL, 0);

	highlight_syntax_in_doc(doc, syndoc, classdoc);

	if (overwrite) {
		xmlSaveFile(fname, doc);
	} else {
		xmlSaveFile("-", doc);
	}

	xmlFreeDoc(doc);
	xmlFreeDoc(syndoc);
	xmlFreeDoc(classdoc);
}

void show_help(void)
{
	puts("Usage: " PROG_NAME " [options] [<dmodule>...]");
	puts("");
	puts("Options:");
	puts("  -h -?      Show usage message.");
	puts("  -c <XML>   Use a custom classes XML file.");
	puts("  -f         Overwrite input data modules.");
	puts("  -s <XML>   Use a custom syntax definitions XML file.");
}

int main(int argc, char **argv)
{
	int i;
	bool overwrite = false;
	char *syntax = NULL;
	char *classes = NULL;

	while ((i = getopt(argc, argv, "c:fs:h?")) != -1) {
		switch (i) {
			case 'c':
				if (!classes)
					classes = strdup(optarg);
				break;
			case 'f':
				overwrite = true;
				break;
			case 's':
				if (!syntax)
					syntax = strdup(optarg);
				break;
			case 'h':
			case '?':
				show_help();
				exit(0);
		}
	}

	if (optind >= argc) {
		highlight_syntax_in_file("-", syntax, classes, false);
	} else {
		for (i = optind; i < argc; ++i) {
			highlight_syntax_in_file(argv[i], syntax, classes, overwrite);
		}
	}

	free(syntax);
	free(classes);

	xmlCleanupParser();

	return 0;
}
